### PR TITLE
Automate workflow testing.

### PR DESCRIPTION
**Pull request by revit13**
This PR automates the process of testing the workflows by implementing the `test` makefile rule in `kfp/transform_workflows` directory.
For the workflow testing one common python virtual env is created. It is expect to be automatically created upon dependency changes. (when the docker image versions are changed for example)
It also loads the docker images to the kind nodes before testing.

### CODE REVIEW

File: kfp/transform_workflows/hack/test_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import os
```
Comment (blublinsky) [inline kfp/transform_workflows/hack/test_pipeline.py]: I think we should move this code to the library

File: kfp/transform_workflows/hack/test_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    """
+    Upload a pipeline. If the pipeline exists delete it and redeploy.
+
+    :param utils: object of type PipelinesUtils
+    :param pipeline_name: Name of the pipeline to be shown in the UI.
+    :param pipeline_package_path: Name of the pipeline to be shown in the UI.
+    return: Server response object containing pipleine id and other information.
+    """
+    pipeline = utils.upload_pipeline(pipeline_package_path=pipeline_package_path, pipeline_name=pipeline_name)
+    if pipeline is None:
```
Comment (blublinsky) [inline kfp/transform_workflows/hack/test_pipeline.py]: I think the simpler thing is to try to get pipeline by name and if it exists delete it before uploading

File: kfp/transform_workflows/hack/test_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    status, error = utils.wait_pipeline_completion(run_id=run_id, timeout=tmout, wait=wait)
+    if status.lower() not in ["succeeded", "completed"]:
+        # Execution failed
+        print(f"Pipeline {pipeline_name} failed with error {error} and status {status}")
+        return None
+    print(f"Pipeline {pipeline_name} successfully completed")
+    return pipeline_name
+
+
+def run_sanity_test(pipeline_package_path: str = "", endpoint: str = "http://localhost:8080/ml-pipeline"):
```
Comment (blublinsky) [inline kfp/transform_workflows/hack/test_pipeline.py]: How is this different from test?

File: kfp/transform_workflows/universal/doc_id/Makefile [hunk 4]
```diff
@@ -1,6 +1,8 @@
-# Define the root of the local git clone for the common rules to be able 
-# know where they are running from.
-REPOROOT=../../../..
+REPOROOT ?= ${CURDIR}/../../../../
+# venv is common to all workflows
+VENV_ACTIVATE ?= ${REPOROOT}/kfp/transform_workflows/venv/bin/activate
+PYTHON ?= python3
```
Comment (blublinsky) [inline kfp/transform_workflows/universal/doc_id/Makefile]: I think it should be python, not python3 as in other places

**Review by blublinsky — COMMENTED** [review]
couple of questions

File: kfp/transform_workflows/hack/test_pipeline.py (hunk 2 above)
Comment (revit13) [reply kfp/transform_workflows/hack/test_pipeline.py]: Fixed. Thanks

File: kfp/transform_workflows/hack/test_pipeline.py (hunk 3 above)
Comment (revit13) [reply kfp/transform_workflows/hack/test_pipeline.py]: Fixed. Thanks

File: kfp/transform_workflows/universal/doc_id/Makefile (hunk 4 above)
Comment (revit13) [reply kfp/transform_workflows/universal/doc_id/Makefile]: Fixed. Thanks

File: kfp/kfp_ray_components/Makefile [hunk 5]
```diff
@@ -31,6 +31,10 @@ reconcile-requirements::
 	sed -i.back "s/kfp-data-processing:[0-9].*/kfp-data-processing:${KFP_DOCKER_VERSION}/" cleanupRayComponent.yaml
 	sed -i.back "s/kfp-data-processing:[0-9].*/kfp-data-processing:${KFP_DOCKER_VERSION}/" executeRayJobComponent_multi_s3.yaml
 
+load-image:
+	@# Help:  Load the image to the kind cluster created with setup-cluster.
```
Comment (roytman) [inline kfp/kfp_ray_components/Makefile]: should it be 
"Load the image to the kind cluster created with `make setup`"?

File: kfp/transform_workflows/Makefile [hunk 6]
```diff
@@ -1,59 +1,44 @@
...
-
 include makeenv
 # Include the common rules.
 # Use "make help" to see them.
 include ../../.make.defaults
 
 include .make.transforms_workflows
 
-PYTHON=python
+export PYTHON=python
```
Comment (roytman) [inline kfp/transform_workflows/Makefile]: should it be `export PYTHON?=python`

File: kfp/transform_workflows/Makefile [hunk 7]
```diff
@@ -1,59 +1,44 @@
...
+export REPOROOT := ${CURDIR}/../..
 include ${REPOROOT}/.make.versions
 
 CURR_DIR := ${CURDIR}
 
 IGNORE := $(shell bash -c "sed -n /=/p  ${REPOROOT}/kfp/requirements.env | sed 's/=/:=/' | sed 's/^/export /' > makeenv")
 
-
 include makeenv
 # Include the common rules.
```
Comment (blublinsky) [inline kfp/transform_workflows/Makefile:4]: Where is makeenv?

File: kfp/transform_workflows/universal/doc_id/Makefile (hunk 4 above)
Comment (blublinsky) [reply kfp/transform_workflows/universal/doc_id/Makefile]: This still shows python3

File: kfp/kfp_support_lib/src/kfp_support/workflow_support/utils/workflow_utils.py [hunk 8]
```diff
@@ -142,6 +142,42 @@ def __init__(self, host: str = "http://localhost:8080"):
         :param host: host to connect to
         """
         self.kfp_client = Client(host=host)
+    
+    def upload_pipeline(
+        self,
+        pipeline_package_path: str = None,
```
Comment (roytman) [inline kfp/kfp_support_lib/src/kfp_support/workflow_support/utils/workflow_utils.py:148]: why it is package? should it be a pipeline yaml file?

File: kfp/transform_workflows/.make.transforms_workflows [hunk 9]
```diff
@@ -1,7 +1,50 @@
...
-.reconcile-requirement::
+define set_env_var
+	$(eval export $(1)=$(2))
+endef
+
+.PHONY: .transforms_workflows.reconcile-requirement
+.transforms_workflows.reconcile-requirement:
 	sed -i.back "s%${DOCKER_IMAGE_NAME}:.*%${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_VERSION}\"%" ${PIPELINE_FILE}
-	sed -i.back "s%kfp-data-processing:.*%kfp-data-processing:${KFP_DOCKER_VERSION}\"%" ${PIPELINE_FILE}
+	sed -i.back "s/kfp-data-processing:.*/kfp-data-processing:${KFP_DOCKER_VERSION}\"/" ${PIPELINE_FILE}
```
Comment (roytman) [inline kfp/transform_workflows/.make.transforms_workflows:17]: Why do you use `/` as a sed command separator here, but' %' is used in the previous line?

File: kfp/transform_workflows/.make.transforms_workflows [hunk 10]
```diff
@@ -1,7 +1,50 @@
...
+		cd ${REPOROOT} && make setup;  \
+	fi
+	cd ${TRANSFORM_SRC} && $(MAKE) image && $(MAKE) load-image
+	sed -i.back 's%imagePullPolicy: ifNotPresent%imagePullPolicy: Always%' ${CURDIR}/${PIPELINE_FILE}
+	cd ${REPOROOT}/kfp/kfp_ray_components && $(MAKE) image && $(MAKE) load-image
+	. ${VENV_ACTIVATE}  && ${PYTHON} ${REPOROOT}/kfp/transform_workflows/hack/test_pipeline.py -p ${CURDIR}/${PIPELINE_FILE}
+
+${VENV_ACTIVATE}: ${REPOROOT}/.make.versions ${REPOROOT}/kfp/requirements.env ${REPOROOT}/kfp/kfp_ray_components/requirements.txt
+	@# Help: Create the virtual environment common to all workflows
+	sed -i.back "s/data-prep-lab-kfp==[0-9].*/data-prep-lab-kfp==${DPL_LIB_KFP_VERSION}/" ${REPOROOT}/kfp/kfp_ray_components/requirements.txt
```
Comment (roytman) [inline kfp/transform_workflows/.make.transforms_workflows]: the same question about sed

File: kfp/transform_workflows/Makefile [hunk 11]
```diff
@@ -1,59 +1,44 @@
 # Define the root of the local git clone for the common rules to be able
 # know where they are running from.
-REPOROOT=../..
+export REPOROOT := ${CURDIR}/../..
```
Comment (roytman) [inline kfp/transform_workflows/Makefile]: Should it be `export` can it be just set? (here and in other places)

File: kfp/transform_workflows/universal/doc_id/Makefile [hunk 12]
```diff
@@ -1,11 +1,26 @@
-# Define the root of the local git clone for the common rules to be able 
-# know where they are running from.
-REPOROOT=../../../..
+REPOROOT ?= ${CURDIR}/../../../../
```
Comment (roytman) [inline kfp/transform_workflows/universal/doc_id/Makefile]: can these 3 lines be in some share place?

**Review by roytman — COMMENTED** [review]
several comments, but probably all the PR should be changed

File: kfp/kfp_support_lib/src/kfp_support/workflow_support/utils/workflow_utils.py (hunk 8 above)
Comment (revit13) [reply kfp/kfp_support_lib/src/kfp_support/workflow_support/utils/workflow_utils.py:148]: it is taken from the kfp function https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/client/client.py#L1388

File: kfp/transform_workflows/.make.transforms_workflows (hunk 9 above)
Comment (revit13) [reply kfp/transform_workflows/.make.transforms_workflows:17]: I will change it to % in all places. Thanks

File: kfp/transform_workflows/Makefile (hunk 7 above)
Comment (revit13) [reply kfp/transform_workflows/Makefile:4]: This part is removed (with additional Makefile cleanup) Thanks

File: kfp/kfp_ray_components/Makefile [hunk 13]
```diff
@@ -31,6 +31,10 @@ reconcile-requirements::
 	sed -i.back "s/kfp-data-processing:[0-9].*/kfp-data-processing:${KFP_DOCKER_VERSION}/" cleanupRayComponent.yaml
```
Comment (roytman) [inline kfp/kfp_ray_components/Makefile:31]: did we decide to change sed commands to "%" ?

File: kfp/kfp_support_lib/Makefile [hunk 14]
```diff
@@ -2,29 +2,24 @@
...
-
-IGNORE := $(shell bash -c "sed -n /=/p  ${REPOROOT}/kfp/requirements.env | sed 's/=/:=/' | sed 's/^/export /' > makeenv")
-
-include makeenv
 # Include the common rules.
 # Use "make help" to see them.
 include ../../.make.defaults
 
-PYTHON=python
 PIP=$(PYTHON) -m pip
```
Comment (roytman) [inline kfp/kfp_support_lib/Makefile]: this command is part of .make. defaults.

File: kind/Makefile [hunk 15]
```diff
@@ -16,6 +16,7 @@ export TOOLS_DIR=${ROOT_DIR}/hack/tools
 export DEPLOY_KUBEFLOW ?= 1
 export USE_KFP_MINIO ?= 1
 export POPULATE_TEST_DATA ?= 1
+KIND_CLUSTER_NAME ?= dataprep
```
Comment (roytman) [inline kind/Makefile:19]: did we define it in .make.defaults?

File: kfp/kfp_ray_components/Makefile (hunk 13 above)
Comment (revit13) [reply kfp/kfp_ray_components/Makefile:31]: Fixed (changed to /") Thanks

File: kfp/kfp_support_lib/Makefile (hunk 14 above)
Comment (revit13) [reply kfp/kfp_support_lib/Makefile]: Thanks removed it

File: kind/Makefile (hunk 15 above)
Comment (revit13) [reply kind/Makefile:19]: yes.

**Review by daw3rd — COMMENTED** [review]
The transform/makefile side of things LGTM

**Review by roytman — APPROVED** [review]
LGTM

**Pull request merged**